The application automatically crashes when I click on the end time option

# Project Information

* Repository: https://github.com/premnirmal/StockTicker
* Project Name: Stock Widget

# Expected behavior
The application should not crash when I click the end time option Rather, all end time options should be properly displayed

# Actual behavior

The application automatically crashes when I click on the end time option 

# How to reproduce

1-[Download ](https://play.google.com/store/apps/details?id=com.github.premnirmal.tickerwidget)the app and open it
2- Go to the setting
3- click on end time 
4- see bug
* Browser/App version:  Current Version 3.0.4
* Operating system: 6.0.1
* Smartphone: Galaxy j7 

# Recording Of The Bug



![20190507_145133](https://user-images.githubusercontent.com/50115498/57294173-adc26f80-707b-11e9-9722-ddeb73d6d7fc.gif)


# My Github Account
https://github.com/facebook-786



# Logcat

<pre><code>E/AndroidRuntime(11193): FATAL EXCEPTION: main

E/AndroidRuntime(11193): Process: com.github.premnirmal.tickerwidget, PID: 11193

E/AndroidRuntime(11193): android.view.InflateException: Binary XML file line #20: Error inflating class android.widget.TimePicker

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.createView(LayoutInflater.java:633)

E/AndroidRuntime(11193): 	at uk.co.chrisjenx.calligraphy.CalligraphyLayoutInflater.onCreateView(SourceFile:164)

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.onCreateView(LayoutInflater.java:682)

E/AndroidRuntime(11193): 	at uk.co.chrisjenx.calligraphy.CalligraphyLayoutInflater.onCreateView(SourceFile:148)

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:741)

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.inflate(LayoutInflater.java:482)

E/AndroidRuntime(11193): 	at uk.co.chrisjenx.calligraphy.CalligraphyLayoutInflater.inflate(SourceFile:60)

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.inflate(LayoutInflater.java:414)

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.inflate(LayoutInflater.java:365)

E/AndroidRuntime(11193): 	at android.app.TimePickerDialog.<init>(TimePickerDialog.java:134)

E/AndroidRuntime(11193): 	at android.app.TimePickerDialog.<init>(TimePickerDialog.java:102)

E/AndroidRuntime(11193): 	at com.github.premnirmal.ticker.settings.SettingsFragment.onDisplayPreferenceDialog(SourceFile:113)

E/AndroidRuntime(11193): 	at androidx.preference.PreferenceManager.showDialog(SourceFile:530)

E/AndroidRuntime(11193): 	at androidx.preference.DialogPreference.onClick(SourceFile:258)

E/AndroidRuntime(11193): 	at androidx.preference.Preference.performClick(SourceFile:1167)

E/AndroidRuntime(11193): 	at androidx.preference.Preference.performClick(SourceFile:1152)

E/AndroidRuntime(11193): 	at androidx.preference.Preference$1.onClick(SourceFile:182)

E/AndroidRuntime(11193): 	at android.view.View.performClick(View.java:5076)

E/AndroidRuntime(11193): 	at android.view.View$PerformClick.run(View.java:20279)

E/AndroidRuntime(11193): 	at android.os.Handler.handleCallback(Handler.java:739)

E/AndroidRuntime(11193): 	at android.os.Handler.dispatchMessage(Handler.java:95)

E/AndroidRuntime(11193): 	at android.os.Looper.loop(Looper.java:135)

E/AndroidRuntime(11193): 	at android.app.ActivityThread.main(ActivityThread.java:5910)

E/AndroidRuntime(11193): 	at java.lang.reflect.Method.invoke(Native Method)

E/AndroidRuntime(11193): 	at java.lang.reflect.Method.invoke(Method.java:372)

E/AndroidRuntime(11193): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1405)

E/AndroidRuntime(11193): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1200)

E/AndroidRuntime(11193): Caused by: java.lang.reflect.InvocationTargetException

E/AndroidRuntime(11193): 	at java.lang.reflect.Constructor.newInstance(Native Method)

E/AndroidRuntime(11193): 	at java.lang.reflect.Constructor.newInstance(Constructor.java:288)

E/AndroidRuntime(11193): 	at android.view.LayoutInflater.createView(LayoutInflater.java:607)

E/AndroidRuntime(11193): 	... 26 more

E/AndroidRuntime(11193): Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void android.widget.TimePicker$TimePickerDelegate.onRtlPropertiesChanged(int)' on a null object reference

E/AndroidRuntime(11193): 	at android.widget.TimePicker.onRtlPropertiesChanged(TimePicker.java:193)

E/AndroidRuntime(11193): 	at android.view.View.resolvePadding(View.java:13757)

E/AndroidRuntime(11193): 	at android.view.ViewGroup.resolvePadding(ViewGroup.java:6344)

E/AndroidRuntime(11193): 	at android.view.View.initializeScrollbarsInternal(View.java:4849)

E/AndroidRuntime(11193): 	at android.view.View.<init>(View.java:4278)

E/AndroidRuntime(11193): 	at android.view.ViewGroup.<init>(ViewGroup.java:503)

E/AndroidRuntime(11193): 	at android.widget.FrameLayout.<init>(FrameLayout.java:115)

E/AndroidRuntime(11193): 	at android.widget.TimePicker.<init>(TimePicker.java:84)

E/AndroidRuntime(11193): 	at android.widget.TimePicker.<init>(TimePicker.java:80)

E/AndroidRuntime(11193): 	at android.widget.TimePicker.<init>(TimePicker.java:76)

E/AndroidRuntime(11193): 	... 29 more
</code></pre>

--- a/app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsFragment.kt
+++ b/app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsFragment.kt
@@ -15,8 +15,10 @@ import android.content.pm.PackageManager
 import android.net.Uri
 import android.os.Build
 import android.os.Bundle
+import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup.MarginLayoutParams
+import android.widget.EditText
 import android.widget.TimePicker
 import androidx.core.app.ActivityCompat
 import androidx.core.content.ContextCompat
@@ -396,7 +398,7 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
   }
 
   private fun needsPermissionGrant(): Boolean {
-    return Build.VERSION.SDK_INT >= 23 && ContextCompat.checkSelfPermission(
+    return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && ContextCompat.checkSelfPermission(
         requireActivity(),
         Manifest.permission.WRITE_EXTERNAL_STORAGE
     ) != PackageManager.PERMISSION_GRANTED
@@ -537,51 +539,61 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
   private fun createTimePickerDialog(pref: TimePreference): Dialog {
     val listener = TimeSelectedListener(
         pref,
-        if (pref.key == AppPreferences.START_TIME) R.string.start_time_updated else R.string.end_time_updated
+        if (pref.key == AppPreferences.START_TIME) R.string.start_time_updated
+        else R.string.end_time_updated
     )
+    if (pref.key == AppPreferences.START_TIME) {
+      val startTime = appPreferences.startTime()
+      pref.lastHour = startTime[0]
+      pref.lastMinute = startTime[1]
+    } else {
+      val endTime = appPreferences.endTime()
+      pref.lastHour = endTime[0]
+      pref.lastMinute = endTime[1]
+    }
     val dialog = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
       TimePickerDialog(
           context, listener,
           pref.lastHour, pref.lastMinute, true
       )
     } else {
-      val timePicker = TimePicker(requireContext()).apply {
-        setIs24HourView(true)
-        currentHour = pref.lastHour
-        currentMinute = pref.lastMinute
-      }
+      // This is to fix a crash on Samsung devices running Android 6.0.1
+      // See https://github.com/premnirmal/StockTicker/issues/80
+      val timeSelectorLayout = LayoutInflater.from(requireContext())
+          .inflate(R.layout.layout_time_selector, null)
+      val hourText = timeSelectorLayout.findViewById<EditText>(R.id.hour_text)
+      val minuteText = timeSelectorLayout.findViewById<EditText>(R.id.minute_text)
+      hourText.setText(pref.lastHour.toString())
+      hourText.setSelection(hourText.text.length)
+      minuteText.setText(pref.lastMinute.toString())
+      minuteText.setSelection(minuteText.text.length)
       AlertDialog.Builder(requireContext())
-          .setView(timePicker)
+          .setView(timeSelectorLayout)
           .setPositiveButton(R.string.ok) { _, _ ->
-            listener.onTimeSet(timePicker, timePicker.currentHour, timePicker.currentMinute)
+            onTimeSet(
+                pref, hourText.text.toString().toInt(), minuteText.text.toString().toInt(),
+                if (pref.key == AppPreferences.START_TIME) R.string.start_time_updated
+                else R.string.end_time_updated
+            )
           }
+          .setNegativeButton(R.string.cancel, null)
           .create()
     }
     dialog.setTitle(
         if (pref.key == AppPreferences.START_TIME) R.string.start_time else R.string.end_time
     )
     return dialog
-}
+  }
 
-inner class TimeSelectedListener(
-  private val preference: TimePreference,
-  private val messageRes: Int
-) : TimePickerDialog.OnTimeSetListener {
-
-  override fun onTimeSet(
-    picker: TimePicker,
-    hourOfDay: Int,
-    minute: Int
+  private fun onTimeSet(
+    preference: TimePreference,
+    lastHour: Int,
+    lastMinute: Int,
+    messageRes: Int
   ) {
-    val lastHour = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
-      picker.hour
-    } else {
-      picker.currentHour
-    }
-    val lastMinute = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
-      picker.minute
-    } else {
-      picker.currentMinute
+    if (lastHour > 23 || lastHour < 0 || lastMinute > 59 || lastMinute < 0) {
+      InAppMessage.showMessage(requireActivity(), R.string.invalid_time, true)
+      return
     }
     val hourString = if (lastHour < 10) "0$lastHour" else lastHour.toString()
     val minuteString = if (lastMinute < 10) "0$lastMinute" else lastMinute.toString()
@@ -599,5 +611,28 @@ inner class TimeSelectedListener(
       InAppMessage.showMessage(requireActivity(), messageRes)
     }
   }
-}
+
+  inner class TimeSelectedListener(
+    private val preference: TimePreference,
+    private val messageRes: Int
+  ) : TimePickerDialog.OnTimeSetListener {
+
+    override fun onTimeSet(
+      picker: TimePicker,
+      hourOfDay: Int,
+      minute: Int
+    ) {
+      val lastHour = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+        picker.hour
+      } else {
+        picker.currentHour
+      }
+      val lastMinute = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+        picker.minute
+      } else {
+        picker.currentMinute
+      }
+      onTimeSet(preference, lastHour, lastMinute, messageRes)
+    }
+  }
 }

--- a/app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsFragment.kt
+++ b/app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsFragment.kt
@@ -564,9 +564,9 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
       val hourText = timeSelectorLayout.findViewById<EditText>(R.id.hour_text)
       val minuteText = timeSelectorLayout.findViewById<EditText>(R.id.minute_text)
       hourText.setText(pref.lastHour.toString())
-      hourText.setSelection(hourText.text.length)
+      hourText.setSelection(hourText.text.toString().length)
       minuteText.setText(pref.lastMinute.toString())
-      minuteText.setSelection(minuteText.text.length)
+      minuteText.setSelection(minuteText.text.toString().length)
       AlertDialog.Builder(requireContext())
           .setView(timeSelectorLayout)
           .setPositiveButton(R.string.ok) { _, _ ->

--- a/app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsFragment.kt
+++ b/app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsFragment.kt
@@ -3,6 +3,8 @@ package com.github.premnirmal.ticker.settings
 import android.Manifest
 import android.annotation.SuppressLint
 import android.app.Activity
+import android.app.AlertDialog
+import android.app.Dialog
 import android.app.TimePickerDialog
 import android.appwidget.AppWidgetManager
 import android.content.Context
@@ -118,23 +120,9 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
 
   override fun onDisplayPreferenceDialog(preference: Preference) {
     when {
-      preference.key == AppPreferences.START_TIME -> {
+      preference.key == AppPreferences.START_TIME || preference.key == AppPreferences.END_TIME-> {
         val pref = preference as TimePreference
-        val dialog =
-          TimePickerDialog(
-              context, TimeSelectedListener(pref, R.string.start_time_updated),
-              pref.lastHour, pref.lastMinute, true
-          )
-        dialog.setTitle(R.string.start_time)
-        dialog.show()
-      }
-      preference.key == AppPreferences.END_TIME -> {
-        val pref = preference as TimePreference
-        val dialog = TimePickerDialog(
-            context, TimeSelectedListener(pref, R.string.end_time_updated),
-            pref.lastHour, pref.lastMinute, true
-        )
-        dialog.setTitle(R.string.end_time)
+        val dialog = createTimePickerDialog(pref)
         dialog.show()
       }
       else -> super.onDisplayPreferenceDialog(preference)
@@ -362,9 +350,10 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
     run {
       val daysPreference = findPreference<MultiSelectListPreference>(AppPreferences.UPDATE_DAYS)
       val selectedDays = appPreferences.updateDaysRaw()
-      daysPreference.summary = appPreferences.updateDays().joinToString {
-        it.getDisplayName(SHORT, Locale.getDefault())
-      }
+      daysPreference.summary = appPreferences.updateDays()
+          .joinToString {
+            it.getDisplayName(SHORT, Locale.getDefault())
+          }
       daysPreference.values = selectedDays
       daysPreference.onPreferenceChangeListener = object : DefaultPreferenceChangeListener() {
         override fun onPreferenceChange(
@@ -373,13 +362,16 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
         ): Boolean {
           val selectedValues = newValue as Set<String>
           if (selectedValues.isEmpty()) {
-            InAppMessage.showMessage(requireActivity(), R.string.days_updated_error_message, error = true)
+            InAppMessage.showMessage(
+                requireActivity(), R.string.days_updated_error_message, error = true
+            )
             return false
           }
           appPreferences.setUpdateDays(selectedValues)
-          daysPreference.summary = appPreferences.updateDays().joinToString {
-            it.getDisplayName(SHORT, Locale.getDefault())
-          }
+          daysPreference.summary = appPreferences.updateDays()
+              .joinToString {
+                it.getDisplayName(SHORT, Locale.getDefault())
+              }
           stocksProvider.schedule()
           InAppMessage.showMessage(requireActivity(), R.string.days_updated_message)
           broadcastUpdateWidget()
@@ -542,33 +534,70 @@ class SettingsFragment : PreferenceFragmentCompat(), ChildFragment,
     super.onActivityResult(requestCode, resultCode, data)
   }
 
-  inner class TimeSelectedListener(
-    private val preference: TimePreference,
-    private val messageRes: Int
-  ) : TimePickerDialog.OnTimeSetListener {
-
-    override fun onTimeSet(
-      picker: TimePicker,
-      hourOfDay: Int,
-      minute: Int
-    ) {
-      val lastHour = picker.currentHour
-      val lastMinute = picker.currentMinute
-      val hourString = if (lastHour < 10) "0$lastHour" else lastHour.toString()
-      val minuteString = if (lastMinute < 10) "0$lastMinute" else lastMinute.toString()
-      val time = "$hourString:$minuteString"
-      val startTimez = appPreferences.timeAsIntArray(time)
-      val endTimez = appPreferences.endTime()
-      if (endTimez[0] == startTimez[0] && endTimez[1] == startTimez[1]) {
-        showDialog(getString(R.string.incorrect_time_update_error))
-      } else {
-        preferences.edit()
-            .putString(preference.key, time)
-            .apply()
-        preference.summary = time
-        stocksProvider.schedule()
-        InAppMessage.showMessage(requireActivity(), messageRes)
+  private fun createTimePickerDialog(pref: TimePreference): Dialog {
+    val listener = TimeSelectedListener(
+        pref,
+        if (pref.key == AppPreferences.START_TIME) R.string.start_time_updated else R.string.end_time_updated
+    )
+    val dialog = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
+      TimePickerDialog(
+          context, listener,
+          pref.lastHour, pref.lastMinute, true
+      )
+    } else {
+      val timePicker = TimePicker(requireContext()).apply {
+        setIs24HourView(true)
+        currentHour = pref.lastHour
+        currentMinute = pref.lastMinute
       }
+      AlertDialog.Builder(requireContext())
+          .setView(timePicker)
+          .setPositiveButton(R.string.ok) { _, _ ->
+            listener.onTimeSet(timePicker, timePicker.currentHour, timePicker.currentMinute)
+          }
+          .create()
+    }
+    dialog.setTitle(
+        if (pref.key == AppPreferences.START_TIME) R.string.start_time else R.string.end_time
+    )
+    return dialog
+}
+
+inner class TimeSelectedListener(
+  private val preference: TimePreference,
+  private val messageRes: Int
+) : TimePickerDialog.OnTimeSetListener {
+
+  override fun onTimeSet(
+    picker: TimePicker,
+    hourOfDay: Int,
+    minute: Int
+  ) {
+    val lastHour = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      picker.hour
+    } else {
+      picker.currentHour
+    }
+    val lastMinute = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      picker.minute
+    } else {
+      picker.currentMinute
+    }
+    val hourString = if (lastHour < 10) "0$lastHour" else lastHour.toString()
+    val minuteString = if (lastMinute < 10) "0$lastMinute" else lastMinute.toString()
+    val time = "$hourString:$minuteString"
+    val startTimez = appPreferences.timeAsIntArray(time)
+    val endTimez = appPreferences.endTime()
+    if (endTimez[0] == startTimez[0] && endTimez[1] == startTimez[1]) {
+      showDialog(getString(R.string.incorrect_time_update_error))
+    } else {
+      preferences.edit()
+          .putString(preference.key, time)
+          .apply()
+      preference.summary = time
+      stocksProvider.schedule()
+      InAppMessage.showMessage(requireActivity(), messageRes)
     }
   }
+}
 }

--- a/app/src/main/res/layout/layout_time_selector.xml
+++ b/app/src/main/res/layout/layout_time_selector.xml
@@ -1,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    android:layout_gravity="center"
+    android:gravity="center"
+    android:orientation="horizontal"
+    android:padding="@dimen/activity_horizontal_margin"
+    >
+
+  <EditText
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:inputType="number"
+      android:id="@+id/hour_text"
+      android:layout_marginEnd="4dp"
+      android:layout_marginRight="4dp"
+      android:maxLength="2"
+      android:hint="HH"
+      />
+
+  <EditText
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:inputType="number"
+      android:layout_marginStart="4dp"
+      android:layout_marginLeft="4dp"
+      android:id="@+id/minute_text"
+      android:maxLength="2"
+      android:hint="MM"
+      />
+
+</LinearLayout>

--- a/app/src/main/res/values-es/strings.xml
+++ b/app/src/main/res/values-es/strings.xml
@@ -106,6 +106,7 @@
   <string name="ok">OK</string>
   <string name="cancel">Cancelar</string>
   <string name="invalid_number">Número inválido</string>
+  <string name="invalid_time">Tiempo inválido</string>
   <string name="add_positions_prompt">¿Quieres añadir empleos para%1$s?</string>
   <string name="already_in_portfolio">%1$s ya está en su cartera</string>
   <string name="added_to_list">%1$s añadido a su cartera</string>

--- a/app/src/main/res/values-es/whats_new.xml
+++ b/app/src/main/res/values-es/whats_new.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <resources>
   <string-array name="whats_new">
-    <item>Actualizaciones de splash</item>
+    <item>Selector de tiempo de fijación para dispositivos Samsung Marshmallow</item>
   </string-array>
 </resources>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -116,6 +116,7 @@
   <string name="ok">OK</string>
   <string name="cancel">Cancel</string>
   <string name="invalid_number">Invalid number</string>
+  <string name="invalid_time">Invalid time</string>
   <string name="add_positions_prompt">Do you want to add positions for %1$s?</string>
   <string name="already_in_portfolio">%1$s is already in your portfolio</string>
   <string name="added_to_list">%1$s added to your portfolio</string>

--- a/app/src/main/res/values/whats_new.xml
+++ b/app/src/main/res/values/whats_new.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <resources>
   <string-array name="whats_new">
-    <item>Splash update</item>
+    <item>Fix time picker for Samsung Marshmallow devices</item>
   </string-array>
 </resources>

--- a/app/version.properties
+++ b/app/version.properties
@@ -1,2 +1,2 @@
-versionName=3.5.3
-versionCode=353
+versionName=3.5.4
+versionCode=354

--- a/app/version.properties
+++ b/app/version.properties
@@ -1,2 +1,2 @@
-versionName=3.5.5
-versionCode=355
+versionName=3.5.6
+versionCode=356

--- a/app/version.properties
+++ b/app/version.properties
@@ -1,2 +1,2 @@
-versionName=3.5.4
-versionCode=354
+versionName=3.5.5
+versionCode=355

